pilot/pkg/networking/plugin/h2sidecar/h2sidecar.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package h2sidecar
+
+import (
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/auth"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/plugin"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pkg/log"
+)
+
+// Plugin instructs sidecars to speak h2 to matching ports.
+type Plugin struct{}
+
+// NewPlugin returns an instance of the h2sidecar plugin.
+func NewPlugin() plugin.Plugin {
+	return Plugin{}
+}
+
+// If an inbound cluster is being built for a port named "http2-h2s", then add
+// a TLS context so that we talk h2s rather than h2c to the workload.
+func (Plugin) OnInboundCluster(in *plugin.InputParams, cluster *xdsapi.Cluster) {
+	if in.Node == nil {
+		return
+	}
+
+	if in.Node.Type != model.SidecarProxy {
+		return
+	}
+
+	if in.ServiceInstance == nil {
+		return
+	}
+
+	if cluster == nil {
+		return
+	}
+
+	if in.Port == nil {
+		return
+	}
+
+	if in.Port.Name != "http2-h2s" {
```

Comment:
I think it might make sense to have this as a prefix, rather than an absolute name. Just in case we need multiple, one day.
```
	if !strings.HasPrefix(in.Port.Name, "http2-h2s"))
```

Reply:
Sure. I think to follow existing semantics, it needs to match exactly "http2-h2s" or have prefix of "http2-h2s-" with a non-empty suffix.

Reply:
Yeah, but, close enough :)